compiler/noirc_evaluator/src/ssa/ssa_gen/context.rs

```diff
@@ -625,16 +625,118 @@ impl<'a> FunctionContext<'a> {
-        }
-
-        self.builder.insert_cast(value, typ)
+        let incoming_type = self.builder.type_of_value(value);
+
+        let result = match (&incoming_type, typ) {
+            // Casting to field is safe
+            (_, NumericType::NativeField) => value,
+            (
+                Type::Numeric(NumericType::Signed { bit_size: incoming_type_size }),
+                NumericType::Signed { bit_size: target_type_size },
+            ) => {
+                match target_type_size.cmp(incoming_type_size) {
+                    std::cmp::Ordering::Less => {
+                        // If target size is smaller, we do a truncation
+                        self.builder.insert_truncate(value, target_type_size, *incoming_type_size)
+                    }
+                    std::cmp::Ordering::Equal => value,
+                    std::cmp::Ordering::Greater => {
+                        // If target size is bigger, we do a sign extension
+                        let value_as_unsigned = self.insert_safe_cast(
+                            value,
+                            NumericType::unsigned(*incoming_type_size),
+                            location,
+                        );
+                        let half_width = self.builder.numeric_constant(
+                            FieldElement::from(2_i128.pow(incoming_type_size - 1)),
+                            NumericType::unsigned(*incoming_type_size),
+                        );
+                        let value_sign =
+                            self.builder.insert_binary(value_as_unsigned, BinaryOp::Lt, half_width);
+                        let patch = self.builder.numeric_constant(
+                            FieldElement::from(
+                                2_i128.pow(target_type_size) - 2_i128.pow(*incoming_type_size),
+                            ),
+                            NumericType::unsigned(target_type_size),
+                        );
+                        let mut sign_not = self.builder.insert_not(value_sign);
+                        sign_not = self.insert_safe_cast(
+                            sign_not,
+                            NumericType::unsigned(target_type_size),
+                            location,
+                        );
+                        // multiplication by a boolean cannot overflow
+                        let patch_with_sign_predicate = self.builder.insert_binary(
+                            patch,
+                            BinaryOp::Mul { unchecked: true },
+                            sign_not,
+                        );
+                        let value_as_unsigned = self.builder.insert_cast(
+                            value_as_unsigned,
+                            NumericType::unsigned(target_type_size),
+                        );
+                        // Patch the bit sign, which gives a `target_type_size` bit size value, so it does not overflow.
+                        self.builder.insert_binary(
```

Comment:
Several instructions here - can you add a comment above the match case maybe on the instructions that are added?
E.g.
```
(for a value v0 of type iN cast to iM)
v1 = cast v0 to uN
v2 = lt v1, 2^(M - 1)
...etc
```

Reply:
Done